mlflow/data/digest_utils.py

```diff
@@ -99,6 +104,27 @@ def compute_tensor_digest(tensor) -> str:
     return compute_numpy_digest(tensor.numpy())
 
 
+def compute_spark_df_digest(df) -> str:
+    """
+    Computes a digest for the given Spark DataFrame. Retrieve a semantic hash of the
+    DataFrame's logical plan, which is much more efficient and deterministic than hashing
+    DataFrame records
+
+    :param df: A Spark DataFrame.
+    :return: A string digest.
+    """
+
+    import pyspark
+    import numpy as np
+
+    # Spark 3.1.0+ has a semanticHash() method on DataFrame
+    if Version(pyspark.__version__) >= Version("3.1.0"):
+        semantic_hash = df.semanticHash()
+    else:
+        semantic_hash = df._jdf.queryExecution().analyzed().semanticHash()
```

Comment:
Have we verified that this method is present in Scala Spark 3.0?

Reply:
Yup, should be there since spark 2.2.0 

see https://issues.apache.org/jira/browse/SPARK-20229 and https://github.com/apache/spark/pull/17541